api/v1alpha1/clusterconfig_types.go

```diff
@@ -344,6 +348,71 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
 	}
 }
 
+type Users []User
+
+func (Users) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Description: "Users to add to the machine",
+			Type:        "array",
+			Items:       ptr.To(User{}.VariableSchema().OpenAPIV3Schema),
+		},
+	}
+}
+
+// User defines the input for a generated user in cloud-init.
+type User struct {
+	// Name specifies the user name
+	Name string `json:"name"`
+
+	// HashedPassword specifies a hashed password for the user. See your
+	// distribution's documentation for instructions to create a hashed
+	// password.
+	// An empty string is not marshalled, because it is not a valid value.
+	// +optional
+	HashedPassword string `json:"hashedPassword,omitempty"`
+
+	// SSHAuthorizedKeys specifies a list of public SSH keys to write to the
+	// machine. Use the corresponding private SSH keys to authenticate. See SSH
+	// documentation for instructions to create a key pair.
+	// +optional
+	SSHAuthorizedKeys []string `json:"sshAuthorizedKeys,omitempty"`
+
+	// Sudo specifies a sudo role for the user
+	// +optional
+	Sudo *string `json:"sudo,omitempty"`
+}
+
+func (User) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
+				"name": {
+					Description: "The username",
+					Type:        "string",
+				},
+				"hashedPassword": {
+					Description: "The hashed password for the user",
+					Type:        "string",
+				},
+				"sshAuthorizedKeys": {
+					Description: "A list of SSH authorized keys for this user",
+					Type:        "array",
+					Items: &clusterv1.JSONSchemaProps{
+						// No description, because the one for the parent array is enough.
+						Type: "string",
+					},
+				},
+				"sudo": {
+					Description: "The sudo rule that applies to this user",
+					Type:        "string",
```

Comment:
I wonder if there is a format/pattern that could be applied here to prevent invalid config? Could be done as a follow-up or not at all...

Reply:
Valid config is definitely desirable, because sudo fails if any rule has invalid syntax, preventing anyone (not just the user that introduced the invalid configuration) from using sudo.

Sudo rules are defined using an EBNF grammar. We would need a parser to validate a rule.

Reply:
That seems overkill 😅 so probably not worth it

Reply:
Added a comment explaining why we don't validate the input